serverless/cmd/feeder/github/main.go

```diff
@@ -0,0 +1,561 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// feeder/github is a feeder implementation for GitHub Action based serverless logs.
+//
+// This tool feeds checkpoints from a github-based serverless log into one or more
+// witnesses, and then raises a PR against the log repo to add any resulting
+//  additional witness signatures to the log's checkpoint.witnessed file.
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-billy/v5/util"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/golang/glog"
+	"github.com/google/go-github/v37/github"
+
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/client"
+	"github.com/google/trillian-examples/serverless/cmd/feeder/impl"
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/oauth2"
+
+	wit_http "github.com/google/trillian-examples/witness/golang/client/http"
+)
+
+// TODO: copied from feed-to-github; adapt as necessary.
+const usage = `Usage:
+ feed-to-github --log_owner_repo --witness_owner_repo --log_repo_path --feeder_config_file --interval
+
+Where:
+ --log_owner_repo is the repo owner/fragment from the repo URL.
+     e.g. github.com/AlCutter/serverless-test -> AlCutter/serverless-test
+ --witness_owner_repo is the repo owner/fragment of the forked log to use for the PR branch.
```

Comment:
`witness_owner_repo` confused me a lot. I think calling it the feeder, distributor, or merger's repo would make more sense. I thought I'd missed something and we'd implemented witnesses which themselves were just github repositories. That aside, it's confusing because this tool supports multiple witnesses, and there's only one repo here.

Reply:
good spot, done